### PR TITLE
Faster RequestServices

### PR DESCRIPTION
**Pull request by benaadams**
Its an oddly chunky middleware

Changes

* `httpContext.Features` to local to reduce virtual lookups from 3 to 1 in common path
* Merge two finally blocks (`using` and `finally`)
* Inlinable `RequestServicesFeature` .ctor 

Pre

Post

I tried using a `ThreadStatic` cache for `RequestServicesFeature` and while it reduced most of the allocations; it was slower than allocating 😭

### CODE REVIEW

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesFeature.cs [hunk 1]
```diff
@@ -24,6 +24,20 @@ public RequestServicesFeature(IServiceScopeFactory scopeFactory)
             _scopeFactory = scopeFactory;
         }
 
+        internal RequestServicesFeature()
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesFeature.cs]: This is an "internal" type. We're ok changing it.

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs [hunk 2]
```diff
@@ -25,38 +25,48 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceScopeFac
...
             }
 
-            _scopeFactory = scopeFactory;
             _next = next;
+            _scopeFactory = scopeFactory;
         }
 
         public async Task Invoke(HttpContext httpContext)
         {
             if (httpContext == null)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: This can't really happen either. I wonder if we should just Debug.Assert here as well. We don't check for a null context in our middleware in general...

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs [hunk 3]
```diff
@@ -25,38 +25,48 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceScopeFac
...
+            var features = httpContext.Features;
+            var existingFeature = features.Get<IServiceProvidersFeature>();
 
-            // All done if request services is set
+            // All done if RequestServices is set
             if (existingFeature?.RequestServices != null)
             {
                 await _next.Invoke(httpContext);
                 return;
             }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: My immediate thought is to care this up into a non-async path and an async path, but my guess is that we'd always take the async path.

Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: Yeah always takes the async path; it does change them into tail call jmps but still ends up doing a little more

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesFeature.cs (hunk 1 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesFeature.cs]: Changed

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs (hunk 2 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: Changed

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs [hunk 4]
```diff
@@ -25,37 +26,37 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceScopeFac
...
 
-            using (var feature = new RequestServicesFeature(_scopeFactory))
+            var replacementFeature = new RequestServicesFeature(_scopeFactory);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: Any reason for removing the using?

Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: It becomes
```csharp
try
{
    try
    {
        // ...
    }
    finally
    {
        feature.Dispose();
    }
}
finally
{
    features.Set(existingFeature);
}
```
So you pay the cost for the double try/finally; but Dispose should never throw so its just merging them

Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: Though with the state machine; I think its 4 try/finallys rather than just 2 😱

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: Because of the async state machine?

Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: State machine adds at least another two `try` blocks; one for execution context restore and one for sticking exception in task (may also be a third - the asm/il is a little hard to follow).

But this is just dealing with the `using` as it creates a `try` block with the `Dispose` in the `finally` and then there is the `try`/`finally` for setting the feature back that's inside the using. So its just turning them into one `try`/`finally` block rather than two.

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: I was being dense. We *already* had an explicit try finally. Sure it makes sense to merge them

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs [hunk 5]
```diff
@@ -25,37 +26,37 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceScopeFac
...
-                try
-                {
-                    httpContext.Features.Set<IServiceProvidersFeature>(feature);
-                    await _next.Invoke(httpContext);
-                }
-                finally
-                {
-                    httpContext.Features.Set(existingFeature);
-                }
+                replacementFeature.Dispose();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: The order here changed from the previous implementation. I'm not sure it's relevant though.

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs (hunk 4 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: 😄

File: src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs (hunk 5 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: Reversed the order as know for certain that `Dispose` doesn't throw; but not sure what features is so it could throw on the `Set`? So it has the same disposal guarantees that the previous arrangement had, even though the execution is different.

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs]: I guess `IFeatureCollection` *is* an abstraction and in theory it could throw. If you were relying on the order of dispose vs set it would probably be broken in other subtle ways... 😄
